src/librustc/hir/lowering.rs

```diff
@@ -4815,6 +4825,18 @@ impl<'a> LoweringContext<'a> {
                                             ThinVec::new()));
         P(self.expr_call(e.span, from_err, hir_vec![e]))
     }
+
+    fn check_tuple_struct_self_ctor_feature(&self, qp: &hir::QPath) {
+        if let hir::QPath::Resolved(_, ref p) = qp {
+            if p.segments.len() == 1 &&
+               p.segments[0].ident.name == keywords::SelfType.name() &&
+               !self.sess.features_untracked().tuple_struct_self_ctor {
+                emit_feature_err(&self.sess.parse_sess, "tuple_struct_self_ctor",
+                                 p.span, GateIssue::Language,
+                                 "tuple struct Self constructors are unstable");
```

Comment:
Nit: this is not about tuple structs, unit structs have `Self` constructors too, so the wording could be tweaked to "`Self` struct constructors are unstable" without mentioning tuples.

Reply:
The feature name `tuple_struct_self_ctor` contains "tuple" too.

Reply:
The original feature name came from the RFC. Now it is renamed to `self_struct_ctor`. And related test files are renamed.